server/api/events/user_message_events.js

```diff
@@ -0,0 +1,81 @@
+var express = require('express');
+var router = express.Router();
+var Moment = require('moment-timezone');
+const ApiValidator = require("./../api_validator");
+const UserMessageEvents = require("../../models/index").UserMessageEvents;
+const Message = require("../../models/index").Message;
+
+const VALIDATOR = new ApiValidator({
+  state: {
+    defaultValue: 0,
+    maxValue: Message.MESSAGE_STATE_VIEWED,
+    minValue: Message.MESSAGE_STATE_UNCHECKED,
+    type: "integer"
+  }
+});
+
+/* GET events related to a speficic user for a specific message. */
+router.get("/user/:userId/message/:messageId", function(req, res) {	
+  res.jsonResultOf(UserMessageEvents.findReadMessageEventsForUser(req.params.userId, req.params.messageId, { deep: true })
+  .then(function(usermesageevents) {
+    if (!usermesageevents) {
+      throw { status: 404 };
+    }
+    // A message may be viewed only by the sender, the receiver, or an admin.
+    if (!req.isAdmin && req.user.id != usermesageevents.fromUserId && req.user.id != usermesageevents.toUserId) {
+      throw { status: 401 };
+    }
+    return usermesageevents;
+  }));
+});
+
+/* GET events related to a speficic user for a specific message. */
+router.get("/allunread", function(req, res) { 
+  res.jsonResultOf(Message.findUnreadMessages({ deep: true })
+  .then(function(messages) {
+    if (!messages) {
+      throw { status: 404 };
+    }
+    // A message may be viewed only by the sender, the receiver, or an admin.
+    if (!req.isAdmin && req.user.id != messages.fromUserId && req.user.id != messages.toUserId) {
+      throw { status: 401 };
+    }
+    return messages;
+  }));
+});
+
+//Create Message Viewed Event
+router.post("/newMessageViewedEvent/:id", function(req, res) {
+  res.jsonResultOf(new Promise(function(resolve) {
+    var fields = VALIDATOR.prevalidateUpdate(req.body);
+    resolve(Message.findById(req.params.id)
+      .then(function(message) {
+        if (!message) {
+          throw { status: 404 };
+        }
+        fields = VALIDATOR.postvalidateUpdate(message, fields);
+        if (!fields) {
+          return message;
+        } else {
+        var userId = (message.toUserId === null) ? req.user.id : message.toUserId ;  
+        return UserMessageEvents.create({
+        type: 'open',
+        clientTime: Date.now(),

```

Comment:
There is a misunderstanding here.  I specified a general event logging endpoint, but this endpoint is specific to a certain type of event.  As we add new event types to the system, we don't want to create new endpoints for them.  There should be a single endpoint that accepts a package of data that describes the event, and takes action based on the contents of the package.